src/SFA.DAS.CandidateAccount.Data/SavedVacancy/SavedVacancyRepository.cs

```diff
@@ -1,37 +1,63 @@
 ﻿using Microsoft.EntityFrameworkCore;
+using Microsoft.IdentityModel.Tokens;
 using SFA.DAS.CandidateAccount.Domain.Candidate;
 
 namespace SFA.DAS.CandidateAccount.Data.SavedVacancy
 {
     public interface ISavedVacancyRepository
     {
-        Task<List<Domain.Candidate.SavedVacancy>> GetByCandidateId(Guid candidateId);
-        Task<Domain.Candidate.SavedVacancy?> Get(Guid candidateId, string vacancyReference);
+        Task<List<Domain.Candidate.SavedVacancy?>> GetByCandidateId(Guid candidateId);
+        Task<Domain.Candidate.SavedVacancy?> Get(Guid candidateId, string? vacancyId, string? vacancyReference);
+        Task<List<Domain.Candidate.SavedVacancy?>> GetAllByVacancyReference(Guid candidateId, string vacancyReference);
         Task<Tuple<Domain.Candidate.SavedVacancy, bool>> Upsert(Domain.Candidate.SavedVacancy savedVacancy);
         Task Delete(Domain.Candidate.SavedVacancy savedVacancy);
     }
 
     public class SavedVacancyRepository(ICandidateAccountDataContext dataContext) : ISavedVacancyRepository
     {
-        public async Task<List<Domain.Candidate.SavedVacancy>> GetByCandidateId(Guid candidateId)
+        public async Task<List<Domain.Candidate.SavedVacancy?>> GetByCandidateId(Guid candidateId)
         {
             var savedVacancyItems = await dataContext.SavedVacancyEntities
                 .AsNoTracking()
                 .Where(x => x.CandidateId == candidateId)
                 .ToListAsync();
 
-            return savedVacancyItems.Select(x => (Domain.Candidate.SavedVacancy)x).ToList();
+            return savedVacancyItems.Select(x => (Domain.Candidate.SavedVacancy?)x).ToList();
         }
 
-        public async Task<Domain.Candidate.SavedVacancy?> Get(Guid candidateId, string vacancyReference)
+        public async Task<Domain.Candidate.SavedVacancy?> Get(Guid candidateId, string? vacancyId, string? vacancyReference)
         {
-            var result = await dataContext.SavedVacancyEntities
+            var query = dataContext.SavedVacancyEntities
                 .AsNoTracking()
-                .SingleOrDefaultAsync(x => x.CandidateId == candidateId && x.VacancyReference == vacancyReference);
+                .Where(x => x.CandidateId == candidateId);
+
+            SavedVacancyEntity? result = null;
+
+            if (!string.IsNullOrEmpty(vacancyId))
+            {
+                query = query.Where(x => x.VacancyId == vacancyId);
+            }
+            else if (!string.IsNullOrEmpty(vacancyReference))
+            {
+                query = query.Where(x => x.VacancyReference == vacancyReference && x.VacancyId == null);
+            }
+
+            result = await query.SingleOrDefaultAsync();
```

Comment:
I feel like we could end up with more than one here?

Reply:
We should never end up with more than one result from this query. Everything calls this query before deciding whether to save a new SavedVacancy. The pre-existing entries have been saved with a unique vacancyReference and null vacancyId. Every save command now saves a unique vacancyId along with the vacancyReference. Null vacancyIds correspond to the first Id for the Reference. And the delete commands from the summary screens now delete all SavedVacancies by the vacancyReference. So the logic means there will only ever be one record for each Reference which may or may not have the vacancyId saved. It recognises that null vacancyIds correspond to the first Id for the Reference.